onnxruntime/core/codegen/common/registry.h

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/common/common.h"
+#include <functional>
+#include <string>
+#include <unordered_map>
+
+namespace onnxruntime {
+namespace codegen {
+
+// RegistryBase is a customized unordered_map
+// that keep ownership of passes,
+// including 1) IR builder passes
+//           2) Weight layout transformer passes
+//           3) Scheduler passses, etc.
+
+template <typename CONTENT_TYPE>
+class RegistryBase {
+ public:
+  RegistryBase() = default;
+
+  bool Contains(const std::string& name) const {
```

Comment:
>const std::string& name [](start = 16, length = 23)

if we hardcode the key type to string, how could you handle the other cases, like op's domain/version? I understand that for initial implementation, we don't need to consider version/domain, but after you go to production, you will hit this issue soon or later.

Reply:
Good suggestion. For this generic class, string could be anything. We can expand the key to include all necessary info.

---
In reply to: [294431551](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294431551) [](ancestors = 294431551)

Reply:
Currently opset and domain is handled inside the op, see codegen/common/target/op_ir_creator/slice.cc as an example. If not critical let's revisit the design in later PRs.

---
In reply to: [294436742](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294436742) [](ancestors = 294436742,294431551)